ostd/src/orpc/oqueue/ringbuffer/mpmc.rs

```diff
@@ -244,33 +263,48 @@ impl<T, const STRONG_OBSERVERS: bool, const WEAK_OBSERVERS: bool>
             core::hint::spin_loop();
         }
         unsafe { slot.store(data) };
-        // Mark the slot as writeable
+        // Mark the slot as readable
         slot.turn.store(self.turn(head, true), Ordering::Release);
     }
 
     fn try_produce(&self, data: T) -> Option<T>
     where
         T: Send,
     {
-        // Get the current head position
-        let mut head = loop {
-            let head = self.head.load(Ordering::Acquire);
-            if head >= MPMCOQUEUE_HEAD_SENTINEL {
-                while self.head.load(Ordering::Relaxed) >= MPMCOQUEUE_HEAD_SENTINEL {
-                    // TODO(aneesh): Revisit the need for yield - this might only be needed in our tests
-                    // that don't have preemptive schedueling.
-                    Task::yield_now();
-                }
-            } else {
-                break head;
-            }
-        };
-
         // We don't need to check for sentinels from this point onwards. If the sentinal value is
         // set, we will just fail below during one of the compare_exchange calls below.
         loop {
+            // Get the current head position
+            let head = loop {
+                let head = self.head.load(Ordering::Acquire);
+                if head >= MPMCOQUEUE_HEAD_SENTINEL {
+                    while self.head.load(Ordering::Relaxed) >= MPMCOQUEUE_HEAD_SENTINEL {
+                        // TODO(aneesh): Revisit the need for yield - this might only be needed in our tests
+                        // that don't have preemptive schedueling.
+                        Task::yield_now();
+                    }
+                } else {
+                    break head;
+                }
+            };
             // Get the slot that we'd like to write to
             let slot = &self.get_slot(head);
+
+            // Check if there are any consumers/observers that can see this message. It's safe to
+            // check this here because if we've successfully gotten a head value to produce to, it's
+            // guaranteed that all active tails are initialized.
+            if self.tail.load(Ordering::Acquire) > head
+                && self
+                    .strong_observer_tails
+                    .iter()
+                    .all(|t| t.load(Ordering::Acquire) > head)
+            {
+                // Mark the slot as writable so that the produce isn't blocked
+                slot.turn
+                    .store(self.next_turn(head, false), Ordering::Release);
+                return None;
```

Comment:
We can't return here. We do need to produce the value for weak observers. We just need to immediately mark it as writable, so that it can be overwritten by a future produce. Could you add/update the tests to check that weak observers behave correctly in the no-strong-observer, no-consumer, and neither cases?

I actually think this is unsafe or at least incorrect since it writes the turn without writing the data in the slot.

Reply:
I think the invariant that turn is never updated unless slot is updated is important. Without it, a weak observer could read an old value thinking it is a new one at best, and read invalid data at worst (though I don't think this is a risk of any code you have written).